DOCS-10593-jms-performance-kt (#1422)
mulesoft/docs-connectors: jms/1.7/modules/ROOT/pages/jms-performance.adoc

= Tune JMS Connector for Performance
:keywords: jms, connector, consume, message, source, listener
:page-aliases: connectors::jms/jms-performance.adoc

> Performance can be a factor when processing the loading of messages. The following recommendations can help you get the most of Anypoint Connector for JMS (JMS Connector) for your Mule applications.

Reviewer: ```suggestion
Performance is sometimes negatively impacted by message loading. The following recommendations can help you get the most of Anypoint Connector for JMS (JMS Connector) for your Mule applications.
```

> == Do not Disable Connections Caching

Reviewer: ```suggestion
== Do Not Disable Connections Caching
```
or
```suggestion
== Don't Disable Connections Caching
```

Reviewer: But...we typically do not want to make titles complete sentences, which these are.
I first thought we could just use a bullet list, but these are too long for that.
Maybe "Connection Caching Reuse"?

> Connections are expensive to create, therefore reuse them as much as you can. By default, JMS Connector uses a caching strategy configuration that reuses as many consumers and producers as possible. Do not disable connection caching because modifying this configuration can reduce the performance of the application.

Reviewer: ```suggestion
Because connections are expensive to create, reuse them as much as you can. By default, JMS Connector uses a caching strategy that reuses as many consumers and producers as possible. To avoid performance degradation, do not disable connection caching.
```

> == Increase Concurrent Messages Processing

Reviewer: Maybe
```suggestion
== Increased Concurrent Messages Processing
```

> To improve the performance of your application you can increase the number of concurrent consumers that receive messages from the same destination. JMS Connector enables you to increase the number of consumers by configuring the *Number of consumers* field in the *On New Message* source. By default, this source uses four consumers concurrently on the same destination, but you can increase this number to the one that fits your needs.

Reviewer: ```suggestion
To improve the performance of your application, increase the number of consumers that concurrently receive messages from the same destination. JMS Connector enables you to increase the number of consumers by configuring the *Number of consumers* field in the *On New Message* source. By default, this source enables four consumers to receive messages concurrently from the same destination, but you can increase this number to the one that best fits your needs.
```

> In the following example, you increase the number of consumers:

Reviewer: ```suggestion
In the following example, you increase the number of consumers receiving messages concurrently:
```

. In Studio, select the *On New Message* source from your flow.
. In the *On New Message* source configuration screen, set the *Number of consumers* field to `20`.

.JMS Connector Number of consumers Configuration
image::jms-performance-1.png[In the On New Message configuration screen, set the Number of consumers field to 20]

In the XML editor, the `numberOfConsumers` configuration looks like this:

[source,xml,linenums]
----
<jms:listener config-ref="config" destination="#[vars.destination]" numberOfConsumers="20"/>
----

> == Optimize the Configuration for Clusters

Reviewer: maybe
```suggestion
== Cluster Configuration Optimization
```

> For applications running in clusters, keep in mind the concept of the primary node and how the connector behaves. When the app runs in a cluster, the *On New Message* source default behavior receives messages only in the primary node, no matter what kind of destination the app is consuming.

Reviewer: "keep in mind" is idiomatic but common. Safe side:
```suggestion
For applications running in clusters, consider the primary node and how the connector behaves. When the app runs in a cluster, the *On New Message* source default behavior receives messages only in the primary node, no matter what kind of destination the app consumes.
```

Reviewer: In some of the following edits, I assumed that the subject of "When consuming <blah>" is the connector. Please fix if not. The point is that the opening phrase "when doing this," must be followed by the thing doing it, to be grammatical (proper placement of modification).
For example, incorrect:
When eating quickly, _choking_ can occur.
Correct:
When eating quickly, _you_ might choke.
Does that make sense?

> In case of consuming messages from a queue, the default behavior is receiving messages in all the cluster's nodes, not just the primary node. JMS Connector enables you to configure the *Primary node only* field in the *On New Message* source. When consuming from a queue in a cluster, do not select the *Primary node only* field, which defaults to `false`.

Reviewer: ```suggestion
When consuming messages from a queue, the default connector behavior is receiving messages in all the cluster's nodes, not just the primary node. JMS Connector enables you to configure the *Primary node only* field in the *On New Message* source. When the connector consumes from a queue in a cluster, do not select the *Primary node only* field, which defaults to `false`.
```

In the following example, you configure the connector to receive messages in all the nodes:

. In Studio, select the *On New Message* source from your flow.

> . In the *Advanced* tab, do not select the *Primary node only* field.

Reviewer: Do not select, or unselect?

.JMS Connector Primary Node only Configuration
image::jms-performance-2.png[In the Advanced tab of the On New Message source, do not select the Primary node only field]

In the XML editor, the `primaryNodeOnly` configuration looks like this:

[source,xml,linenums]
----
<jms:listener config-ref="config" destination="${inputQueue}" primaryNodeOnly="false"/>
----

> In case of consuming messages from a topic, the default behavior is receiving messages only in the primary node, which avoids processing the same message multiple times across the cluster. +

Reviewer: ```suggestion
When consuming messages from a topic, the default connector behavior is to receive messages only in the primary node, which avoids processing the same message multiple times across the cluster. +
```

> When consuming from a topic in a cluster, not selecting the the *Primary node only* field causes the cluster to process the same message more than once, unless you also configure shared-subscriptions. +

Reviewer: ```suggestion
When consuming messages from a topic in a cluster, a connector with *Primary node only* unselected causes the cluster to process the same message more than once, unless you also configure shared subscriptions. +
```

If you are using the JMS 2.0 shared-subscriptions mechanism, then change the cluster configuration to consume from all the nodes, again setting he *Primary node only* field to `false`.

> In the following example, you configure the connector using 2.0 spec shared-subscriptions mechanism:

Reviewer: ```suggestion
In the following example, you configure the connector using the shared-subscriptions mechanism:
```

. In Studio, select the *On New Message* source from your flow.
. In the *On New Message* source configuration screen, select *Topic consumer*.
. Select the *Shared* option.

> . Set the *Subscription name* to `clusteredEventListener`

Reviewer: ```suggestion
. Set the *Subscription name* to `clusteredEventListener`.
```

> . In the *Advanced* tab, do not select the *Primary node only* field.

Reviewer: do not select or unselect? "leave unselected"?

.JMS Connector Primary Node only Configuration
image::jms-performance-3.png[In the Advanced tab of the On New Message source, do not select the Primary node only field]

> In the XML editor, the `jms:topic-consumer`, `shared` and `primaryNodeOnly` configurations look like this:

Reviewer: ```suggestion
In the XML editor, the `jms:topic-consumer`, `shared`, and `primaryNodeOnly` configurations look like this:
```
